### ISSUE TITLE
(v1.1.289) langserver.index.js does not exist

### ISSUE THREAD

**Issue by FrancescElies** [RobertCraigie/pyright-python#138]
I get that error by just calling the following.

```python
❯ pyright-langserver.exe --help

up to date, audited 2 packages in 778ms

found 0 vulnerabilities
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\s\myrepo\venv\py311\Scripts\pyright-langserver.exe\__main__.py", line 7, in <module>
  File "C:\Users\FrancescElies\src\pyright-python\pyright\langserver.py", line 57, in entrypoint
    sys.exit(main(*sys.argv[1:]))
             ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\FrancescElies\src\pyright-python\pyright\langserver.py", line 26, in main
    return run(*args, **kwargs).returncode
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\FrancescElies\src\pyright-python\pyright\langserver.py", line 51, in run
    raise RuntimeError(f'Expected language server entrypoint: {binary} to exist')
RuntimeError: Expected language server entrypoint: C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\langserver.index.js to exist
```

I see node output which tried to install node on the temp path above but that folder is empty.
This corresponds to the execution of this [line](https://github.com/RobertCraigie/pyright-python/blob/main/pyright/langserver.py#L47) of code

```
up to date, audited 2 packages in 778ms

found 0 vulnerabilities
```
Despite of npm install the binary does [not exist](https://github.com/RobertCraigie/pyright-python/blob/main/pyright/langserver.py#L50-L51)

What am I doing wrong?

```
❯ npm --version
8.14.0
```

**Comment by RobertCraigie** [comment]
I can't reproduce this locally on MacOS.

Could you please share what is contained in that directory? It may also be helpful to run with [debug logging](https://github.com/RobertCraigie/pyright-python#debugging) enabled and share the output here too.

**Comment by FrancescElies** [comment]
```
pyright-python on  main [⇣] via 🐍 v3.11.1 took 8s
❯ let-env PYRIGHT_PYTHON_DEBUG = 1

pyright-python on  main [⇣] via 🐍 v3.11.1
❯ pyright-langserver.exe --help
2023-01-18 12:49:10,329 - DEBUG - pyright.node - Checking for global target binary: npm
2023-01-18 12:49:10,329 - DEBUG - pyright.node - Found global binary at: C:\Program Files\nodejs\npm.CMD
2023-01-18 12:49:10,329 - DEBUG - pyright.node - Global binary exists at: C:\Program Files\nodejs\npm.CMD
2023-01-18 12:49:10,329 - DEBUG - pyright.node - Running node command with args: ['C:\\Program Files\\nodejs\\npm.CMD', 'install', 'pyright@1.1.289']

up to date, audited 2 packages in 617ms

found 0 vulnerabilities
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\s\eklang\venv\py311\Scripts\pyright-langserver.exe\__main__.py", line 7, in <module>
  File "C:\Users\FrancescElies\src\pyright-python\pyright\langserver.py", line 57, in entrypoint
    sys.exit(main(*sys.argv[1:]))
             ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\FrancescElies\src\pyright-python\pyright\langserver.py", line 26, in main
    return run(*args, **kwargs).returncode
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\FrancescElies\src\pyright-python\pyright\langserver.py", line 51, in run
    raise RuntimeError(f'Expected language server entrypoint: {binary} to exist')
RuntimeError: Expected language server entrypoint: C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\langserver.index.js to exist
```

**Comment by FrancescElies** [comment]
So the issue is that this command `['C:\\Program Files\\nodejs\\npm.CMD', 'install', 'pyright@1.1.289']` despite of being executed in the temp folder it installs pyright in this location `~\node_modules\pyright\langserver.index.js`

Not sure if this is a platform specific thing of npm in windows.

What do you think, what shall we do? If you feel it helps I am happy to provide a PR once we decide how this should be addressed.

**Comment by FrancescElies** [comment]
The previous pr seems to help now I can

```
pyright-python on  main [!?⇣] is 📦 v1.0.0 via  v19.1.0 via 🐍 v3.11.1
❯ pyright-langserver.exe --help
2023-01-18 13:11:30,696 - DEBUG - pyright.node - Checking for global target binary: node
2023-01-18 13:11:30,712 - DEBUG - pyright.node - Found global binary at: C:\Program Files\nodejs\node.EXE
2023-01-18 13:11:30,712 - DEBUG - pyright.node - Global binary exists at: C:\Program Files\nodejs\node.EXE
2023-01-18 13:11:30,712 - DEBUG - pyright.node - Running node command with args: ['C:\\Program Files\\nodejs\\node.EXE', 'C:\\Users\\FRANCE~1\\AppData\\Local\\Temp\\pyright-python-langserver.FrancescElies\\node_modules\\pyright\\langserver.index.js', '--', '--help']
Error: Connection input stream is not set. Use arguments of createConnection or set command line parameters: '--node-ipc', '--stdio' or '--socket={number}'
    at _createConnection (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\pyright-internal\node_modules\vscode-languageserver\lib\node\main.js:189:15)
    at createConnection (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\pyright-internal\node_modules\vscode-languageserver\lib\node\main.js:136:12)
    at run (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\pyright-internal\src\nodeServer.ts:22:19)
    at main (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\pyright-internal\src\nodeMain.ts:14:5)
    at Object.call (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\pyright\src\langserver.ts:3:1)
    at __webpack_require__ (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\webpack\bootstrap:19:32)
    at fn (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\webpack\bootstrap:35:79)
    at Function.O (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\webpack\runtime\chunk loaded:23:12)
    at next (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\webpack\bootstrap:36:44)
    at Function.s.x (C:\Users\FRANCE~1\AppData\Local\Temp\pyright-python-langserver.FrancescElies\node_modules\pyright\dist\webpack\runtime\startup chunk dependencies:5:9)

pyright-python on  cesc/fix-executables-does-not-exist via 🐍 v3.11.1
❯ cat c:\s\mytest\pyright_join_iter_literalstring.py | pyright-langserver.exe --stdio
2023-01-18 13:11:38,068 - DEBUG - pyright.node - Checking for global target binary: node
2023-01-18 13:11:38,084 - DEBUG - pyright.node - Found global binary at: C:\Program Files\nodejs\node.EXE
2023-01-18 13:11:38,084 - DEBUG - pyright.node - Global binary exists at: C:\Program Files\nodejs\node.EXE
2023-01-18 13:11:38,084 - DEBUG - pyright.node - Running node command with args: ['C:\\Program Files\\nodejs\\node.EXE', 'C:\\Users\\FRANCE~1\\AppData\\Local\\Temp\\pyright-python-langserver.FrancescElies\\node_modules\\pyright\\langserver.index.js', '--', '--stdio']
Content-Length: 119

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Pyright language server 1.1.289 starting"}}Content-Length: 213

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":3,"message":"Server root directory: C:\\Users\\FRANCE~1\\AppData\\Local\\Temp\\pyright-python-langserver.FrancescElies\\node_modules\\pyright/dist/"}}
```

If that pr is the right fix I am not sure, don't know much about npm

**Comment by RobertCraigie** [comment]
Ah, thank you for the PR!

Yeah that seems like the right fix to me, I have ran into this issue in a separate project before as I had a file present at `~/package.json` which meant that any `npm install`s in directories such as `~/.cache/` resulted in packages being downloaded to `~/node_modules`. I also ran into that issue on Windows so I'm not sure if it's an issue that will only occur on Windows or not but it's worth fixing regardless.

Thanks for figuring this out! I'll try to merge the PR today.
